File: algo/src/main/java/com/hartwig/actin/algo/evaluation/priortumor/HasActiveSecondMalignancy.java
package com.hartwig.actin.algo.evaluation.priortumor;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.datamodel.ImmutableEvaluation;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;
import com.hartwig.actin.clinical.datamodel.PriorSecondPrimary;

import org.jetbrains.annotations.NotNull;

public class HasActiveSecondMalignancy implements EvaluationFunction {

    HasActiveSecondMalignancy() {
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        boolean hasMatch = false;
        for (PriorSecondPrimary priorSecondPrimary : record.clinical().priorSecondPrimaries()) {
            if (priorSecondPrimary.isActive()) {
                hasMatch = true;
            }
        }

        EvaluationResult result = hasMatch ? EvaluationResult.PASS : EvaluationResult.FAIL;
        ImmutableEvaluation.Builder builder = ImmutableEvaluation.builder().result(result);
        if (result == EvaluationResult.FAIL) {
            builder.addFailSpecificMessages("Patient has no active second malignancy");
        } else if (result == EvaluationResult.PASS) {
            builder.addPassSpecificMessages("Patient has active second malignancy");
        }

        return builder.build();
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/priortumor/HasActiveSecondMalignancy.kt
package com.hartwig.actin.algo.evaluation.priortumor

import com.hartwig.actin.algo.evaluation.EvaluationFactory
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.Evaluation
import com.hartwig.actin.datamodel.clinical.TumorStatus

class HasActiveSecondMalignancy : EvaluationFunction {

    override fun evaluate(record: PatientRecord): Evaluation {
        return if (record.priorPrimaries.any { it.status == TumorStatus.ACTIVE }) {
            EvaluationFactory.pass("Active second malignancy in provided history")
        } else if (record.priorPrimaries.any { it.status == TumorStatus.EXPECTATIVE }) {
            EvaluationFactory.warn("Second malignancy with expectative status in provided history")
        } else {
            EvaluationFactory.fail("No active second malignancy")
        }
    }
}